### DISCUSSION TITLE
Passing vars to routes

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#344] [Q&A] [answered]
Where please is it documented how that should be implemented with axum?

From a buildable example I made: (https://github.com/kellytk/x-variables/blob/main/src/main.rs#L11)

`let app = Router::new().route("/", get(endpoint::handler(var)));`

If that's impossible, then is https://docs.rs/axum/0.2.5/axum/#sharing-state-with-handlers the only method available?

**Comment by 3olkin** [comment]
In axum handlers accept zero or several extractors. Extractor is a type that implement FromRequest trait ([more info](https://docs.rs/axum/0.2.5/axum/extract/index.html)).
So you have several ways to pass variable to handler:
1. Use shared state
2. Manually implement FromRequest for your type
3. Use lazy_static crate or similar for some static vars like env vars (JWT_SECRET and similar). You can think about it as global constants which will load only once.

**Comment by jplatte** [comment, marked as answer]
From the example it kind of looks like you want something like this:

```rust
let app = Router::new().route("/", get(move |body: Json<Something>| {
    let var = var.clone();
    endpoint::handler(body, var)
}));
```

but the `AddExtension` middleware would probably easier to work with.

**Reply by kellytk** [reply]
@jplatte Thank you.  Both that and the `AddExtension` method feel heavy compared with Warp which affords syntax akin to what I wrote originally. (using [BoxedFilter](https://docs.rs/warp/0.3.1/warp/filters/struct.BoxedFilter.html))

May I ask what the design rationale or advantage is over passing arguments directly to handlers?

FWIW, my cause here is capturing env vars once for customizing program behavior on start up.  Passing that via a request facility feels like overloading it.

**Reply by jplatte** [reply]
I don't get what you mean at all. I don't see the connection between the link you posted to the original question.

**Reply by kellytk** [reply]
[Warp-based buildable example](https://github.com/kellytk/x-warpvars/blob/main/src/main.rs#L10)

**Reply by davidpdrsn** [reply]
Thats just the design I chose for axum: To use function arguments as extractors. Its very convenient for some things and less convenient for others. Just trade-offs really. Warp has an entirely different design with different trade-offs. What jplatte posted is the way to go.

**Reply by kellytk** [reply]
Understood, thank you.
